odxtools/codec.py

```diff
@@ -0,0 +1,212 @@
+# SPDX-License-Identifier: MIT
+import typing
+from typing import TYPE_CHECKING, List, Optional, runtime_checkable
+
+from .decodestate import DecodeState
+from .encodestate import EncodeState
+from .exceptions import EncodeError, odxraise
+from .odxtypes import ParameterValue
+
+if TYPE_CHECKING:
+    from .parameters.parameter import Parameter
+
+
+@runtime_checkable
+class Codec(typing.Protocol):
+    """Any object which can be en- or decoded to be transferred over
+    the wire implements this API.
+    """
+
+    @property
+    def short_name(self) -> str:
+        return ""
+
+    def encode_into_pdu(self, physical_value: Optional[ParameterValue],
+                        encode_state: EncodeState) -> None:
+        ...
+
+    def decode_from_pdu(self, decode_state: DecodeState) -> ParameterValue:
+        ...
+
+    def get_static_bit_length(self) -> Optional[int]:
+        ...
+
+
+@runtime_checkable
+class CompositeCodec(Codec, typing.Protocol):
+    """Any object which can be en- or decoded to be transferred over
+    the wire which is composed of multiple parameter implements this
+    API.
+
+    """
+
+    @property
+    def parameters(self) -> List["Parameter"]:
```

Comment:
I am not 100% sure, but I think this is more correct 
```suggestion
    def parameters(self) -> NamedItemList[Parameter]:
```

Reply:
I think it should stay at `List`, because `NamedItemList` is a `List` and we should probably not rely on any `NamedItemList` specifics when we assume a `Codec` object...

Reply:
ok, remove the extra quotes then

Reply:
Not sure if this is a good idea: `Parameter` is a `CompositeCodec`, so whilst this would work in the present version of the PR (but not in the original one!), importing `Parameter` in `codec.py`  would IMO be something like a "cyclic import by spirit". (this sounds quite esoteric :D. in any case, mypy works as expected because of the conditional import of `Parameter` at the beginning of the file.)

Reply:
ok, changed my mind. (see below). probably I should factor out the helper functions from `codec.py` into `codec_helpers.py`, but that's work for another day...